docs/DesignDocs/ConfigIDInterface.md

```diff
@@ -0,0 +1,209 @@
+Verifier Plugin might support other integrity algorithms. If the caller of the
+`oe_result_t oe_get_evidence(...)` function sets an integrity algorithm ID not
+supported by the Verifier Plugin, the Verifier Plugin should output the
+`inittime_custom_claims_buffer` as it is, as an unverified init-time claim. In
+that case, the consumer of the outputted claims is responsible to verify the
+integrity of the `inittime_custom_claims_buffer` claim using the base claim of
+config_id/config_svn.
+
+```C
+/**
+ * oe_get_evidence
+ *
+ * Generates the evidence for the given format id.
+ * This function is only available in the enclave.
+ *
+ * @param[in] format_id The format ID of the evidence to be generated.
+ * @param[in] flags A bit-wise parameter. Currently there is one bit
+ * defined: OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID. If this bit is set,
+ * the evidence and endorsements will be wrapped with a header containing
+ * the format ID.
+ * @param[in] runtime_custom_claims_buffer The optional runtime custom claims
+ * buffer. When provided, the content of the buffer is included in the
+ * evidence_buffer, with integrity protection. Depending on the underlining TEE
+ * and the plugin implementation, the content might or might not be further
+ * encrypted.
+ * @param[in] runtime_custom_claims_buffer_size The number of bytes in the
+ * runtime custom claims buffer.
+ * @param[in] inittime_custom_claims_buffer The optional inittime custom claims
+ * buffer. When provided, the content of the buffer is included in the
+ * evidence_buffer as plaintext. The integrity protection of the content is
+ * provided by the underlining TEE and the enclave SW outside the plugin.
+ * @param[in] inittime_custom_claims_buffer_size The number of bytes in the
+ * inittime custom claims buffer.
+ * @param[in] optional_parameters The optional format-specific input parameters.
+ * @param[in] optional_parameters_size The size of optional_parameters in bytes.
+ * @param[out] evidence_buffer An output pointer that will be assigned the
+ * address of the dynamically allocated evidence buffer.
+ * @param[out] evidence_buffer_size A pointer that points to the size of the
+ * evidence buffer in bytes.
+ * @param[out] endorsements_buffer If not NULL, an output pointer that will be
+ * assigned the address of the dynamically allocated endorsements buffer.
+ * @param[out] endorsements_buffer_size A pointer that points to the size of the
+ * endorsements buffer in bytes.
+ * @retval OE_OK The function succeeded.
+ * @retval OE_INVALID_PARAMETER At least one of the parameters is invalid.
+ * @retval OE_NOT_FOUND The input evidence format id is not supported.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_get_evidence(
+    const oe_uuid_t* format_id,
+    uint32_t flags,
+    const void* runtime_custom_claims_buffer,
+    size_t runtime_custom_claims_buffer_size,
+    const oe_inittime_custom_claim_buffer_t* inittime_custom_claims_buffer,
```

Comment:
It's unclear to me why this needs to be a separate set of parameters in the `oe_get_evidence` function itself:

- From the attester side, if the desire is simply to bind an extra buffer to the resulting evidence blob, then including it as part of the existing `custom_claims_buffer` would be sufficient.
   - If syntactic sugar is desired for structuring the buffer from name value pairs, it would be preferable to shift that functionality into the helper functions like `oe_serialize_custom_claims`
- If the intent is to imply special tagging of the information for handling by an enlightened verifier plugin in the implementation of `oe_verify_evidence`, it seems like the verification of the integrity of a custom claim against information in the evidence can be better generalized across TEEs by pushing it into a separate function instead tightly coupling it to basic evidence structure verification performed by `oe_verify_evidence`. 
    - In the parlance of the IETF RATS aligned [AttestationApiRequirements.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/AttestationApiRequirements.md), this might be something in the `appraise_evidence`/`set_evidence_appraisal_policy` implementation on the verifier side.

Reply:
The caller of the `oe_get_evidence` API has different expectations on how the attester plugin handles the run-time claims and  init-time claims. The plugin is supposed to include the run-time claim in the evidence in a way that provide integrity protection of the run-time claims. For example, the SGX Attester plugin sets the `reportdata` field of SGX REPORT/QUOTE as SHA256 hash of the run-time claims. The plugin is supposed to simply include the init-time claims in the evidence and leave the integrity protection responsibility to the underlining TEE and enclave SW outside of the plugin. For example, the SGX enclave developer implements enclave code that bind the init-time claim buffer to SGX ConfigID/ConfigSVN. To delineate the run-time claims vs. init-time claims, either the `oe_get_evidence` API supports different input parameters for run-time claims and init-time claims or the `oe_serialize_custom_claims` API does that. I think either approach can work. But if `oe_serialize_custom_claims` API syntax is considered TEE or plugin-implementation specific, the caller would have to write different calling code for different `oe_serialize_custom_claims` API syntax. As `oe_get_evidence` API syntax is considered TEE agonistic and plugin implementation agonistic, adding the dedicated init-time claims related input parameters in `oe_get_evidence` API enables the caller to have the same calling code for different TEEs and plugin implementations that support init-time claims concept.

The other part of the proposal in this PR - including an init-time claims integrity-protection algorithm identifier in the init-time claim buffer, is to enable enlightened verifier plugins to verify the integrity of the init-time claims included in the evidence. The goal of the API syntax design is again to enable the caller of the attester side APIs to have the same calling code for different TEES and plugin implementations. The proposal in this PR defines the init-time claims buffer passed to `oe_get_evidence` API as a structure containing the integrity-protection algorithm identifier. Alternatively, a dedicated init-time claims integrity-protection algorithm identifier input parameter can be added to `oe_get_evidence`, or to `oe_serialized_custom_claims`, if `oe_serialized_custom_claims` API is defined as TEE agonistic and plugin implementation agonistic. The verifier side policy about whether `oe_verify_evidence` should verify the integrity of the init-time claims is independent of the the attester side API syntax proposal. For example, instead of `oe_verify_evidence` verifies the integrity of the init-time claims, `oe_verifies_evidence` only outputs unverified init-time claims, and `appraise_evidence` verifies the init_time claims against the Config_ID base claim. But the requirement is that the init-time claims integrity algorithm identifier is provided to the verifier side, within the attestation evidence produced by `oe_get_evidence`.

Reply:
Based on the discussion in the Attestation arch SIG meeting, my understanding is that `oe_serialized_customer_claims` is supposed to be implemented by the enclave developer. In that case, the custom_claims_buffer's internal structure, for example, to delineate the init-time claims and run-time claims, is opaque to the Attester and Verifier Plugin. The Attester Plugin won't be able to locate the run-time claims portion of the custom_claims_buffer to create the integrity-protection metadata (in SGX Attester Plugin, it's SGX `REPORT.report_data`). If the Attester creates the integrity-protection metadata over the whole custom_claim_buffer, the operation only introduces unnecessary extra latency to operate on the init_time claims, with no security benefit.  More importantly, the Verifier Plugin will not be able to locate the run-time claims buffer to verify against the Config_ID.

Pushing the responsibility to verify init-time claims out of `oe_verifies_evidence`/`appraise_evidence` might be OK. But the upper level logic, for example, an Attestation Verification Service, such as [MAA](https://docs.microsoft.com/en-us/azure/attestation/overview) will need to handle that complexity.

Reply:
[PR#3726](https://github.com/openenclave/openenclave/pull/3726) discussed the alternative of leaving the packaging and verification of the customer_claims_buffer to the upper level SW outside of the Attester/Verifier Plugin.